File: Zig/day01/src/fuel.zig
const math = @import("std").math;

fn transform(n: i64) ?i64 {
    return switch (@divFloor(n, 3) - 2) {
        1...math.maxInt(i64) => |num| num,
        else => null,
    };
}

pub const FuelIterator = struct {
    fuel: ?i64,

    const Self = @This();

    pub fn new(fuel: i64) Self {
        return Self { .fuel = fuel };
    }

    pub fn next(self: *Self) ?i64 {
        const fuel = self.fuel orelse return null;
        self.fuel = transform(fuel);
        return self.fuel;
    }
};

File: Rust/day01/src/fuel.rs
pub trait FuelTransform: Sized {
    fn calculate(&self) -> Option<Self>;
}

impl FuelTransform for i64 {
    fn calculate(&self) -> Option<Self> {
        match (self / 3) - 2 {
            n if n > 0 => Some(n),
            _ => None,
        }
    }
}

pub trait Fuel: Copy {
    fn fuel(&self) -> FuelIterator<Self> {
        FuelIterator { fuel: Some(*self) }
    }
}

impl<T: Copy> Fuel for T {}

pub struct FuelIterator<N> {
    fuel: Option<N>,
}

impl<N: FuelTransform + Copy> Iterator for FuelIterator<N> {
    type Item = N;

    fn next(&mut self) -> Option<Self::Item> {
        match self.fuel {
            None => None,
            Some(fuel) => {
                self.fuel = fuel.calculate();
                self.fuel
            }
        }
    }
}
